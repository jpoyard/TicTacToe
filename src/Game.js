import React from 'react';
import { Board } from './Board';
import { calculateWinner } from './calculateWinner';

export class Game extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      history: [{ squares: Array(9).fill(null), move: null }],
      xIsNext: true,
      stepNumber: 0,
      reverseOrder: false,
    };
  }

  handleClick(i) {
    const history = this.state.history.slice(0, this.state.stepNumber + 1);
    const current = history[history.length - 1];
    const squares = current.squares.slice();
    if (calculateWinner(squares) || squares[i]) {
      return;
    }
    squares[i] = this.state.xIsNext ? 'X' : 'O';
    const column = i % 3;
    const row = (i - column) / 3;
    this.setState({
      history: history.concat([
        {
          squares: squares,
          move: { i, row, column },
        },
      ]),
      stepNumber: history.length,
      xIsNext: !this.state.xIsNext,
      reverseOrder: this.state.reverseOrder,
    });
  }

  toggleOrder() {
    this.setState({ reverseOrder: !this.state.reverseOrder });
  }

  jumpTo(step) {
    this.setState({
      stepNumber: step,
      xIsNext: step % 2 === 0,
    });
  }

  render() {
    const history = this.state.history;
    const current = history[this.state.stepNumber];
    const result = calculateWinner(current.squares);
    let moves = history.map((step, move) => {
      const desc = move
        ? 'Revenir au tour n°' +
          move +
          ' {row: ' +
          step.move.row +
          ', column: ' +
          step.move.column +
          '}'
        : 'Revenir au début de la partie';
      return (
        <li
          key={move}
          className={move === this.state.stepNumber ? 'active' : ''}
        >
          <button onClick={() => this.jumpTo(move)}>{desc}</button>
        </li>
      );
    });
    moves = this.state.reverseOrder ? moves.reverse() : moves;
    let status = result
      ? result.winner + ' a gagné'
      : this.state.stepNumber === 9
      ? 'Match null'
      : 'Prochain joueur : ' + (this.state.xIsNext ? 'X' : 'O');

    return (
      <div className="game">
        <div className="game-board">
          <Board
            squares={current.squares}
            winnerSquares={result?.squares ?? []}
            onClick={(i) => this.handleClick(i)}
          />
        </div>
        <div className="game-info">
          <div className="status">{status}</div>
          <button
            className="toggleHistoryBtn"
            onClick={() => this.toggleOrder()}
          >
            Order
          </button>
          <ol>{moves}</ol>
        </div>
      </div>
    );
  }
}
