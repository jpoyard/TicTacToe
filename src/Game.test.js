/* eslint-disable testing-library/no-node-access */
import { cleanup, render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { Game } from './Game';

// eslint-disable-next-line jest/valid-title
describe(Game.name, () => {
  let container = null;
  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    cleanup(container);
    container.remove();
    container = null;
  });

  [
    {
      when: [0, 0, 8, 6, 3, 2, 4, 1, 7],
      then: {
        board: ['X', 'X', 'X', 'O', 'O', '', 'X', '', 'O'],
        status: 'X a gagné',
      },
    },
    {
      when: [0, 1, 8, 4, 7, 6, 3, 2],
      then: {
        board: ['X', 'O', 'O', 'X', 'O', '', 'O', 'X', 'X'],
        status: 'O a gagné',
      },
    },
    {
      when: [0, 4, 6, 3, 5, 8, 2, 1, 7],
      then: {
        board: ['X', 'O', 'X', 'O', 'O', 'X', 'X', 'X', 'O'],
        status: 'Match null',
      },
    },
  ].forEach(({ when, then }) => {
    it(`should manage game correctly when ${then.status}`, () => {
      // Given
      // When
      playGame(container, when);
      // Then
      expect(
        Array.from(document.querySelectorAll('.square')).map(
          (e) => e.textContent
        )
      ).toEqual(then.board);
      expect(document.querySelector('.game-info .status').textContent).toBe(
        then.status
      );
    });
  });

  it('should manage display history correctly', () => {
    playGame(container, [0, 8, 6, 3, 2, 4, 1]);

    const expectedHistory = [
      'Revenir au début de la partie',
      'Revenir au tour n°1 {row: 0, column: 0}',
      'Revenir au tour n°2 {row: 2, column: 2}',
      'Revenir au tour n°3 {row: 2, column: 0}',
      'Revenir au tour n°4 {row: 1, column: 0}',
      'Revenir au tour n°5 {row: 0, column: 2}',
      'Revenir au tour n°6 {row: 1, column: 1}',
      'Revenir au tour n°7 {row: 0, column: 1}',
    ];
    // Check history
    expect(
      Array.from(document.querySelectorAll('li')).map((e) => e.textContent)
    ).toEqual(expectedHistory);

    // Reverse Order
    act(() => {
      document
        .querySelector('button.toggleHistoryBtn')
        .dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(
      Array.from(document.querySelectorAll('li')).map((e) => e.textContent)
    ).toEqual(expectedHistory.reverse());
  });

  it('should manage history jump correctly', () => {
    playGame(container, [0, 8, 6, 3, 2, 4, 1]);
    act(() => {
      document
        .querySelectorAll('li button')[4]
        .dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    expect(
      Array.from(document.querySelectorAll('.square')).map((e) => e.textContent)
    ).toEqual(['X', '', '', 'O', '', '', 'X', '', 'O']);
  });
});

function playGame(container, sequence) {
  const actual = render(<Game />, container).container;
  sequence.forEach((squareIndex) => {
    act(() => {
      document
        .querySelectorAll('.square')
        [squareIndex].dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
  });
  return actual;
}
