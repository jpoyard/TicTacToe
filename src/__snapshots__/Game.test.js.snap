// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should manage display history correctly 1`] = `
<div>
  <div
    class="game"
  >
    <div
      class="game-board"
    >
      <div>
        <div
          class="board-row"
        >
          <button
            class="square win"
          >
            X
          </button>
          <button
            class="square win"
          >
            X
          </button>
          <button
            class="square win"
          >
            X
          </button>
        </div>
        <div
          class="board-row"
        >
          <button
            class="square"
          >
            O
          </button>
          <button
            class="square"
          >
            O
          </button>
          <button
            class="square"
          />
        </div>
        <div
          class="board-row"
        >
          <button
            class="square"
          >
            X
          </button>
          <button
            class="square"
          />
          <button
            class="square"
          >
            O
          </button>
        </div>
      </div>
    </div>
    <div
      class="game-info"
    >
      <div
        class="status"
      >
        X a gagné
      </div>
      <button
        class="toggleHistoryBtn"
      >
        Order
      </button>
      <ol>
        <li
          class=""
        >
          <button>
            Revenir au début de la partie
          </button>
        </li>
        <li
          class=""
        >
          <button>
            Revenir au tour n°1 {row: 0, column: 0}
          </button>
        </li>
        <li
          class=""
        >
          <button>
            Revenir au tour n°2 {row: 2, column: 2}
          </button>
        </li>
        <li
          class=""
        >
          <button>
            Revenir au tour n°3 {row: 2, column: 0}
          </button>
        </li>
        <li
          class=""
        >
          <button>
            Revenir au tour n°4 {row: 1, column: 0}
          </button>
        </li>
        <li
          class=""
        >
          <button>
            Revenir au tour n°5 {row: 0, column: 2}
          </button>
        </li>
        <li
          class=""
        >
          <button>
            Revenir au tour n°6 {row: 1, column: 1}
          </button>
        </li>
        <li
          class="active"
        >
          <button>
            Revenir au tour n°7 {row: 0, column: 1}
          </button>
        </li>
      </ol>
    </div>
  </div>
</div>
`;
